src/mlpack/methods/ann/layer/gru_impl.hpp

```diff
@@ -0,0 +1,322 @@
+      hiddenStateModule);
+
+  forwardStep++;
+  if (forwardStep == rho)
+  {
+    forwardStep = 0;
+    if (!deterministic)
+    {
+      outParameter.push_back(std::move(arma::mat(allZeros.memptr(),
+        allZeros.n_rows, allZeros.n_cols, false, true)));
+      prevOutput = --outParameter.end();
+    }
+    else
+    {
+      *prevOutput = std::move(arma::mat(allZeros.memptr(),
+        allZeros.n_rows, allZeros.n_cols, false, true));
+    }
+  }
+  else if (!deterministic)
+  {
+    outParameter.push_back(output);
+    prevOutput = --outParameter.end();
+  }
+  else
+  {
+    if (forwardStep == 1)
+    {
+      outParameter.clear();
+      outParameter.push_back(output);
+
+      prevOutput = outParameter.begin();
+    }
+    else
+    {
+      *prevOutput = output;
+    }
+  }
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void GRU<InputDataType, OutputDataType>::Backward(
+  const arma::Mat<eT>&& /* input */, arma::Mat<eT>&& gy, arma::Mat<eT>&& g)
+{
+  if ((outParameter.size() - backwardStep  - 1) % rho != 0 && backwardStep != 0)
+  {
+    gy += boost::apply_visitor(deltaVisitor, output2GateModule);
+  }
+
+  if (backIterator == outParameter.end())
+  {
+    backIterator = --(--outParameter.end());
+  }
+
+  // Delta zt.
+  arma::mat dZt = gy % (*backIterator -
+      boost::apply_visitor(outputParameterVisitor,
+      hiddenStateModule));
+
+  // Delta ot.
+  arma::mat dOt = gy % (arma::ones<arma::vec>(outSize) -
+      boost::apply_visitor(outputParameterVisitor, inputGateModule));
+
+  // Delta of input gate.
+  boost::apply_visitor(BackwardVisitor(std::move(boost::apply_visitor(
+      outputParameterVisitor, inputGateModule)), std::move(dZt),
+      std::move(boost::apply_visitor(deltaVisitor, inputGateModule))),
+      inputGateModule);
+
+  // Delta of hidden gate.
+  boost::apply_visitor(BackwardVisitor(std::move(boost::apply_visitor(
+      outputParameterVisitor, hiddenStateModule)), std::move(dOt),
+      std::move(boost::apply_visitor(deltaVisitor, hiddenStateModule))),
+      hiddenStateModule);
+
+  // Delta of outputHidden2GateModule.
+  boost::apply_visitor(BackwardVisitor(std::move(boost::apply_visitor(
+      outputParameterVisitor, outputHidden2GateModule)),
+      std::move(boost::apply_visitor(deltaVisitor, hiddenStateModule)),
+      std::move(boost::apply_visitor(deltaVisitor, outputHidden2GateModule))),
+      outputHidden2GateModule);
+
+  // Delta rt.
+  arma::mat dRt = boost::apply_visitor(deltaVisitor, outputHidden2GateModule) %
+      *backIterator;
+
+  // Delta of forget gate.
+  boost::apply_visitor(BackwardVisitor(std::move(boost::apply_visitor(
+      outputParameterVisitor, forgetGateModule)), std::move(dRt),
+      std::move(boost::apply_visitor(deltaVisitor, forgetGateModule))),
+      forgetGateModule);
+
+  // Put delta zt.
+  prevError.submat(0, 0, 1 * outSize - 1, 0) = boost::apply_visitor(
```

Comment:
That works if we use a batchSize = 1, do you mind to refactor the parameter so that it works with batchSize > 1? You could get the batchSize by looking at ``input.n_cols``, so instead of using ``prevError = arma::zeros<arma::mat>(3 * outSize, 1);`` we have to use ``prevError.set_size(4 * outSize, batchSize);`` and ``prevError.submat(outSize, 0,  2 * outSize - 1, batchSize - 1)``. I can refactor the LSTM layer, but it would be great if you could update the GRU layer.